8277204: Implement PAC-RET branch protection on Linux/AArch64

a74nh opened the pull request:
PAC is an optional feature in AArch64 8.3 and is compulsory in v9. One
of its uses is to protect against ROP based attacks. This is done by
signing the Link Register whenever it is stored on the stack, and
authenticating the value when it is loaded back from the stack. If an
attacker were to try to change control flow by editing the stack then
the authentication check of the Link Register will fail, causing a
segfault when the function returns.

On a system with PAC enabled, it is expected that all applications will
be compiled with ROP protection. Fedora 33 and upwards already provide
this. By compiling for ARMv8.0, GCC and LLVM will only use the set of
PAC instructions that exist in the NOP space - on hardware without PAC,
these instructions act as NOPs, allowing backward compatibility for
negligible performance cost (2 NOPs per non-leaf function).

Hardware is currently limited to the Apple M1 MacBooks. All testing has
been done within a Fedora Docker image. A run of SpecJVM showed no
difference to that of noise - which was surprising.

The most important part of this patch is simply compiling using branch
protection provided by GCC/LLVM. This protects all C++ code from being
used in ROP attacks, removing all static ROP gadgets from use.

The remainder of the patch adds ROP protection to runtime generated
code, in both stubs and compiled Java code. Attacks here are much harder
as ROP gadgets must be found dynamically at runtime. If/when AOT
compilation is added to JDK, then all stubs and compiled Java will be
susceptible ROP gadgets being found by static analysis and therefore
potentially as vulnerable as C++ code.

There are a number of places where the VM changes control flow by
rewriting the stack or otherwise. I’ve done some analysis as to how
these could also be used for attacks (which I didn’t want to post here).
These areas can be protected ensuring the pointers to various stubs and
entry points are stored in memory as signed pointers. These changes are
simple to make (they can be reduced to a type change in common code and
a few addition sign/auth calls in the backend), but there a lot of them
and the total code change is fairly large. I’m happy to provide a few
work in progress patches.

In order to match the security benefits of the Apple Arm64e ABI across
the whole of JDK, then all the changes mentioned above would be
required.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8277204](https://bugs.openjdk.java.net/browse/JDK-8277204): Implement PAC-RET branch protection on Linux/AArch64
 * [JDK-8277543](https://bugs.openjdk.java.net/browse/JDK-8277543): Implement PAC-RET branch protection on Linux/AArch64 (**CSR**)


### Reviewers
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to e0e3f666d85f5b0414b200fb9288928a181875fb
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to 29471d30f2f5b328eaef7c2e2042f65cfbfc698d
 * [Andrew Dinn](https://openjdk.java.net/census#adinn) (@adinn - **Reviewer**) ⚠️ Review applies to 2c27eb5e60c59ec4cd55e4d3805bde07d01ccb6e
 * [Nick Gasson](https://openjdk.java.net/census#ngasson) (@nick-arm - **Reviewer**) ⚠️ Review applies to deb17a566e2405eb98a1b23aec167bb7b2599f0f


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6334/head:pull/6334` \
`$ git checkout pull/6334`

Update a local copy of the PR: \
`$ git checkout pull/6334` \
`$ git pull https://git.openjdk.java.net/jdk pull/6334/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6334`

View PR using the GUI difftool: \
`$ git pr show -t 6334`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6334.diff">https://git.openjdk.java.net/jdk/pull/6334.diff</a>

</details>

theRealAph commented:
Gosh. This is going to take some time to review, and will need at least two reviewers.

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -5154,3 +5162,75 @@ void MacroAssembler::verify_cross_modify_fence_not_required() {
+
+// ROP Protection
+
+void MacroAssembler::protect_return_address() {
```
We need proper, full, detailed comments about what these functions do, with reference to primary AArch64 documentation.

adinn replied:
As far as the AArch64 docs are concerned the relevant details are provided in ARM-ARM D

- The PAC functionality is described in ARM-ARM Section D5.1.5
- Overview of the PAC instructions is provided in section C3.1.9
- Detailed PAC instruction descriptions are provided in C6.2.195 - C6.2.199

n.b. I am specifically referring to my (possibly out of date) copy ARM-DDI 0487D.a (ID103018) which is the Initial v8.4 EAC release from 2018.

That said, I agree that a description of how these functions use the underlying PAC support and what, effectively, they achieve via that usage would be necessary. A reference to the relevant sections of the ARM doc in the code would be helpful.

adinn replied:
Correction:
Using the most up to date ARM ARM G  [ARM DDI 0487G.a (ID011921)]

- The PAC functionality is described in ARM-ARM Section D5.1.5
- Overview of the PAC instructions is provided in section C3.1.10
- Detailed PAC instruction descriptions are provided in C6.2.208 - C6.2.212

a74nh replied:
I'm thinking for references to the Arm Arm to use header titles instead of section numbers, as the titles should be more stable.

Also probably need some description around the code in the pauth_aarch64.hpp too. But I want to make sure I'm not duplicating comments - maybe the macroassembler comments should point to the pauth_aarch64 comments.

It didn't seen common in the code to describe instruction functionality, which is why I didn't add any. Agreed it needs something added though.

theRealAph replied:
Yeah. At the definitions of `authenticate_return_address()` et al you can say what you expect in the normal case and what you expect when you've been hacked, along with an overview. I realize that it was a bit tricky to make this work with HotSpot because we're synthesizing return addresses just like hackers do, so a comment where we're patching return addresses would be nice.

As long as the instructions are easily findable in the docs that's good.

theRealAph replied:
Just to be clear: no, don't describe instructions. describe what the macros do, and when to use them. Imagine that you, the reader can't see the contents of the macro at all, just the name and the comments.

erikj79 reviewed (approved):
Build change looks good, but I can't comment on the code changes.

a74nh commented:
> Gosh. This is going to take some time to review, and will need at least two reviewers.

Sure. And thanks in advance.

theRealAph commented on src/hotspot/os_cpu/bsd_aarch64/pauth_bsd_aarch64.inline.hpp:
```diff
@@ -25,6 +25,8 @@
 #ifndef OS_CPU_BSD_AARCH64_PAUTH_BSD_AARCH64_INLINE_HPP
```
Are these two files different enough to separate them for BSD and Linux?

a74nh replied:
My motivation was to avoid having any ifdefs - but we need one anyway for the apple ifdef.

If I merged the two we would end up with just the contents of the BSD version of the file.

There is also the windows version of the file, which for now has empty functions. If PAC in windows is added, that'll either use the same code or maybe Windows will provide an API (like the Apple one). Merging everything would mean windows gains the UseROPProtection check.

a74nh replied:
>Are these two files different enough to separate them for BSD and Linux?

Merging these files then broke everything for windows (because the asm function is different). Having a "ifdef apple, elseif windows else" doesn't really make sense, so I'll split the files out again.

magicus commented on make/autoconf/flags-cflags.m4:
```diff
@@ -881,3 +883,30 @@ AC_DEFUN([FLAGS_SETUP_GCC6_COMPILER_FLAGS],
+    BRANCH_PROTECTION_AVAILABLE=false
+  elif test "x$TOOLCHAIN_TYPE" = xgcc || test "x$TOOLCHAIN_TYPE" = xclang; then
+    # Check that the compiler actually supports branch protection.
+    FLAGS_COMPILER_CHECK_ARGUMENTS(ARGUMENT: [${BRANCH_PROTECTION_FLAG}],
```
This branch misses a AC_MSG_RESULT, which prints the newline. The resulting output will look messy.

a74nh replied:
Looking at this block of code again, I've got far too many outputted lines compared to other features. Removing some means I can simplify the code too, so I'll do that.

adinn commented:
I am also reviewing this.

magicus reviewed (approved):
Build changes look much better now, thanks!

Build part approved; the actual code changes needs approval from others.

fweimer-rh commented on src/hotspot/cpu/aarch64/globals_aarch64.hpp:
```diff
@@ -110,7 +110,9 @@ define_pd_global(intx, InlineSmallCode,          1000);
-          range(-1, 4096)
+          range(-1, 4096)                                               \
+  product(bool, UseROPProtection, false,                                \
+          "Use ROP based branch protection")                            \
```
The description is not correct. It's protection against certain ROP-based attack techniques.

adinn replied:
I don't agree that this is incorrect, at least not for the stated reason. The flag switches on a protection mechanism that guards against ROP attacks. To my reading that does not imply it guards against all such attacks, merely that this is the nature of the protection it offers.

The description might still be considered incorrect for an unrelated reason. Its use of the adjectival phrase ROP based constitutes a transferred epithet, conflating the symptom with the medicine. In other words, the protection offered is not ROP based i.e. does not rely on an ROP technique. What it does is protect against ROP attacks. So, I'd suggest rewording to

    "Enable protection of branches against ROP attacks".

Florian, if you want to argue for rewording that to "Enable protection of branches against some categories of ROP attacks" or some other equivalently qualified variant please feel free to make a case. However, I don't think see any need to add that rider, nor any precedent in any of the other short descriptions provided in globals.hpp.

fweimer-rh replied:
I did mean the description, not the flag name.

adinn replied:
Yes, understood. I too was talking about the description even though I introduced my comment by talking about what the flag does.

theRealAph replied:
`"Protect branches against ROP attacks".`

adinn commented:
> Is the code still mapped read-write all the time?

That depends on what code you mean. The JVM code compiled from C++ sources is mapped RO(X) in the text section like any compiled C/C++ code. Protection of that code is covered by the changes to the build system.

The runtime generated runtime stubs and Java method code  into which this patch may insert the required PAC instructions are written into a code cache in a section which is mapped RW(X) all the time. It would be hard to map even a subset of this code cache RO because generated code includes call and data sites that need to be patched during execution.

a74nh commented:
> The runtime generated runtime stubs and Java method code into which this patch may insert the required PAC instructions are written into a code cache in a section which is mapped RW(X) all the time. It would be hard to map even a subset of this code cache RO because generated code includes call and data sites that need to be patched during execution.

Am I right is saying that for Macos, all generated code is remapped RO before execution?

An additional concern I have is that if the globals data was attacked then the UseROPProtection flag could be flipped, and all code after that point would be generated without ROP protection. Marking all the globals data as RO would fix that. Alternatively remove UseROPProtection and then in the macroassembler always generate PAC code, using just the subset of instructions that are NOPs on non-PAC hardware. Or alternatively only generate PAC code based on a #define set at build time. Each option has its own downsides.

adinn commented:
> Am I right is saying that for Macos, all generated code is remapped RO before execution?

Ah, no, it seems the code cache is not RWX all the time as far as Java threads are concerned. The Macos/AArch64 code is strategically calling  pthread_jit_write_protect_np at Java <-> JVM transition points.

That ensures that executable regions are executable but not writable (RX) from a Java thread when running JITted Java code and are writable but not executable (RW) when it calls into JVM code.

> An additional concern I have is that if the globals data was attacked then the UseROPProtection flag could be flipped, and all code after that point would be generated without ROP protection. Marking all the globals data as RO would fix that. Alternatively remove UseROPProtection and then in the macroassembler always generate PAC code, using just the subset of instructions that are NOPs on non-PAC hardware. Or alternatively only generate PAC code based on a #define set at build time. Each option has its own downsides.

Globals data can legitimately be written during JVM startup (perhaps in some cases also during execution?). So, they cannot simply be marked as RO.

I am not sure this concern is really warranted. If an attacker is already able to overwrite UseROPProtection then a concern over the resulting omission of JITted ROP protection seems like attending to the loud banging of the stable door while Shergar has already been diced into stew meat.

theRealAph commented:
> > Am I right is saying that for Macos, all generated code is remapped RO before execution?
> 
> Ah, no, it seems the code cache is not RWX all the time as far as Java threads are concerned. The Macos/AArch64 code is strategically calling pthread_jit_write_protect_np at Java <-> JVM transition points.

And this requires magic kernel support. I did mention it to a kernel engineer who wasn't very impressed, but I think it's pretty cool.

fweimer-rh commented:
> > > Am I right is saying that for Macos, all generated code is remapped RO before execution?
> > 
> > 
> > Ah, no, it seems the code cache is not RWX all the time as far as Java threads are concerned. The Macos/AArch64 code is strategically calling pthread_jit_write_protect_np at Java <-> JVM transition points.
> 
> And this requires magic kernel support. I did mention it to a kernel engineer who wasn't very impressed, but I think it's pretty cool.

It's possible to emulate this to some extent with memory protection keys on POWER and (recent) x86. See `pkey_alloc`.

a74nh commented:
*Updated UseROPProtection message
*Moved pauth functions into single file
*Added comments
*Removed superfluous modifier arg from macroassembler funcs

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -5154,3 +5162,98 @@ void MacroAssembler::verify_cross_modify_fence_not_required() {
+// Use directly after authentication to catch authentication failures.
+// Also use before signing to check that the pointer is valid and hasn't already been signed.
+//
+void MacroAssembler::check_return_address(Register return_reg) {
```
This commentary is excellent. Thanks.

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -5163,6 +5163,23 @@ void MacroAssembler::verify_cross_modify_fence_not_required() {
+        ShouldNotReachHere();
+    }
+  }
+}
```
This merge seems to have gone wrong. Please check to make sure that the `OnSpinWait` changes aren't part of your PR.

adinn commented on src/hotspot/cpu/aarch64/pauth_aarch64.hpp:
```diff
@@ -25,11 +25,121 @@
+#endif
 
+// Support for ROP Protection in VM code.
+// This is provided by via the AArch64 PAC feature.
```
"by via" should just be "via"

theRealAph commented on src/hotspot/cpu/aarch64/c1_Runtime1_aarch64.cpp:
```diff
@@ -443,13 +445,12 @@ OopMapSet* Runtime1::generate_handle_exception(StubID id, StubAssembler *sasm) {
+  __ mov(lr, r0);
 
   // only r0 is valid at this time, all other registers have been destroyed by the runtime call
   __ invalidate_registers(false, true, true, true, true, true);
```
Not so: `lr` is live.

theRealAph commented on src/hotspot/cpu/aarch64/c1_Runtime1_aarch64.cpp:
```diff
@@ -443,13 +445,12 @@ OopMapSet* Runtime1::generate_handle_exception(StubID id, StubAssembler *sasm) {
 
-  // patch the return address, this stub will directly return to the exception handler
-  __ str(r0, Address(rfp, 1*BytesPerWord));
-
```
Please explain the reason for this change, that leaves `lr` live across `restore_live_registers()`.

a74nh replied:
In the original code:
*save r0 to the lr location on the stack
*restore_live_registers
*Standard return: remove stack frame, load lr and fp off the stack, jump to lr.
 
With PAC it would now be:
*Sign r0 then save it to the lr location on the stack
*restore_live_registers
*Standard return: remove stack frame, load lr and fp off the stack, auth lr, jump to lr.

After reading the code in restore_live_registers, it doesn't touch lr and so seemed odd to have the save to the stack, only to restore it directly afterwards.

theRealAph replied:
That's an optimization, though. You shouldn't need to read the code in `restore_live_registers()` to see if it's safe to keep the return address in LR: at best it's pathological coupling, in the sense that the correctness of this code depends on the internal details of  `restore_live_registers()`. Let's keep LR live ranges as short as possible.

a74nh replied:
Ok, that's fine, I'll update it (It'll simplify the total code diff too).

theRealAph commented on src/hotspot/cpu/aarch64/pauth_aarch64.hpp:
So, whether this function authenticates or strips the address depends only on debugging? The vague name makes the callers hard to read.

a74nh replied:
>whether this function authenticates or strips the address depends only on debugging?

Yes. We only need to strip the value, because we're not jumping to the lr value, only viewing it.

The interface is different to a strip (as we need to pass in the modifier). 

How about something like pauth_authenticate_fast() ? or pauth_authenticate_unsafe() ?

Alternatively, this function is only called by the functions in Frame, so the frequency of use is probably low enough (compared to the sign/auth every function) that it's not going to cause any performance issues. So, could just replace with calls to pauth_authenticate. I think that might be the best option.

theRealAph replied:
A simple rule here: function names go with what the release version does. So I'd go with the actual purpose, which is `pauth_strip_addr_for_debuginfo()`. That's right, isn't it? You only want this thing for stack traces, logs, etc.

adinn replied:
This function is used by the frame code. So, that means it is used for all stack walks which are far from being simply cosmetic/ornamental. The runtime will rely on this for various different types of thread housekeeping.

The difference here is that in product mode this simply strips auth bits whereas in debug mode it actually authenticates as it strips to give extra verification. So, your suggested name is quite misleading. Likewise Alan's suggested names is misleading because the primary product operation is to strip not authenticate.

How about pauth_strip_verifiable? and a comment saying that it differs from pauth_strip by actually authenticating when debug is enabled.

adinn reviewed (approved):
This is much clearer and looks good to push modulo a minor typo I noted in a comment.

nick-arm commented:
We're adding a product option `UseROPProtection` which needs a CSR according to https://wiki.openjdk.java.net/display/HotSpot/Hotspot+Command-line+Flags%3A+Kinds%2C+Lifecycle+and+the+CSR+Process

nick-arm commented:
/csr needed

a74nh commented:
CSR added: https://bugs.openjdk.java.net/browse/JDK-8277543

a74nh commented:
@dholmes-ora 

Fixed flags based on comments on the CSR:

> However, the proposed implementation does not match the description. VM_Version::initialize() is called after argument processing so even if the user has explicitly set the flag to false, if this is a PAC enabled system it will be turned back on. The code needs to check if the flag has been set before overriding the value; further the warning at line 389 should only be given if the user actually turned the flag on. But this should be taken up in the PR.

dholmes-ora commented:
> @dholmes-ora
> 
> Fixed flags based on comments on the CSR:

Flag updates look good - thanks.

theRealAph commented on src/hotspot/cpu/aarch64/vm_version_aarch64.cpp:
```diff
@@ -408,6 +408,25 @@ void VM_Version::initialize() {
+  if (_features & CPU_PACA) {
+    if (FLAG_IS_DEFAULT(UseROPProtection)) {
+      FLAG_SET_DEFAULT(UseROPProtection, true);
+    }
```
Please leave this off by default, at least for now.

theRealAph commented on make/autoconf/flags-cflags.m4:
```diff
@@ -881,3 +883,24 @@ AC_DEFUN([FLAGS_SETUP_GCC6_COMPILER_FLAGS],
+
+  BRANCH_PROTECTION_CFLAGS=""
+  UTIL_ARG_ENABLE(NAME: branch-protection, DEFAULT: auto,
+      RESULT: USE_BRANCH_PROTECTION, AVAILABLE: $BRANCH_PROTECTION_AVAILABLE,
```
What exactly is going on here? Is it that if the host compiler has "branch protection" supported, we will build with it on by default? if so, no, we don't want to do that. For now, branch protection should be an explicit opt in.

a74nh replied:
The idea was that the same JDK could be used on PAC and non PAC systems. I agree that right now it's probably better being explicit opt in. Will update.

theRealAph commented on src/hotspot/cpu/aarch64/vm_version_aarch64.cpp:
```diff
@@ -408,6 +408,25 @@ void VM_Version::initialize() {
+  } else {
+    if (UseROPProtection) {
+      warning("UseROPProtection specified, but not supported on this CPU.");
+      FLAG_SET_DEFAULT(UseROPProtection, false);
```
```suggestion
      FLAG_SET_DEFAULT(UseROPProtection, true);
```
Given that the instructions used are in NOP space, this won't do any harm, and it will allow developers without PAC-enabled systems to see what code PAC would generate.

a74nh replied:
Ok, I think that's fine. How about on a non pac system allowing it for development only ?

theRealAph replied:
Maybe. Mind you, a lot of the time I'm looking at the output from production systems.
From a rather philosophical point of view, I assume that if the user of a computer asks for something that isn't going to break anything or confuse anyone, we should honour their request.

theRealAph replied:
Was this ever resolved?

a74nh replied:
Sort of. That code has changed quite a bit - UseROPProtection now is a string field not a bool.
"none" or not set - pac disabled.
"pac-ret" - pac always enabled
"standard" - pac enabled if the cpu+os support it.

Also, the pac instructions used aren't all in the NOP space. So, it will crash on a non-pac machine.  It might be possible to change it so it does only use nop space instructions, but I don't think it'll be optimal (need to double check).

theRealAph replied:
OK, cool.

theRealAph commented on src/hotspot/cpu/aarch64/globals_aarch64.hpp:
```diff
@@ -118,7 +118,9 @@ define_pd_global(intx, InlineSmallCode,          1000);
           "It cannot be used with OnSpinWaitInst=none.")                \
-          range(1, 99)
+          range(1, 99)                                                  \
+  product(bool, UseROPProtection, false,                                \
```
Question: this is called "UseROPProtection", the configure option is called "enable-branch-protection", and GCC option is called "-mbranch-protection". This is confusing. I would have thought we would want the same name, and use it for all branch protection. So why is this not "UseBranchProtection"?

fweimer-rh replied:
`mbranch-protection` switches on both PAC-RET and BTI. This PR only covers a use of PAC that looks very ROP-focused to me.

theRealAph replied:
True, because we don't (yet) support BTI. Is there any point having two separate flags for BTI and PAC-RET? If someone wants one, they'll very likely want the other, won't they?

a74nh replied:
You can support one without the other.
The architecture allows you to have one without the other.
The GCC flag is an enum of "none|standard|pac-ret[+leaf]|bti", with some of them changing depending on which cpu you specify to -mcpu (8.0,8.3,8.5 etc).
Clang has the same flags. Interestingly, on MacOS Clang, -mbranch-protection is available but it'll give incorrect code. Instead you build with -arch arm64e.

If your system had both, the only scenario I could see for only wanting just one would be for test/dev purposes. In a real production scenario you would want everything the system supports or nothing.

An earlier version of my code had a UseBranchProtection="pac|bti|pac+bti|all|none" style option

theRealAph replied:
> You can support one without the other. The architecture allows you to have one without the other. The GCC flag is an enum of "none|standard|pac-ret[+leaf]|bti", with some of them changing depending on which cpu you specify to -mcpu (8.0,8.3,8.5 etc). Clang has the same flags. 

OK, so we have a precedent.

> If your system had both, the only scenario I could see for only wanting just one would be for test/dev purposes. In a real production scenario you would want everything the system supports or nothing.

Yes.

> An earlier version of my code had a UseBranchProtection="pac|bti|pac+bti|all|none" style option

That sounds great.

It seems to me that following the GCC/Clang precedent is the wisest thing we could do. I can see no possible advantage in diverging: it only confuses programmers.

a74nh replied:
That gives us:

A new flag -XX:UseBranchProtection

With the options:
none - no PAC support. (Default)
standard - PAC support if the system supports it and the java binary was compiled with PAC. Otherwise off.
pac-ret - PAC support, regardless if the system supports it or the java binary was compiled with PAC.

A later BTI patch would add:
standard - also adds BTI if the system supports it and the java binary was compiled with BTI.
bti - BTI support, regardless if the system supports it or the java binary was compiled with BTI.
Also, concat the flags with "+". Eg: standard+bti. No need to do this until BTI is added.


For MacOS, you can only use PAC functionality when compiled for arm64e. Therefore arm64e would be supported by compiling the java binary for the arm64e and would always be enabled in that scenario. UseBranchProtection on MacoOS will only support the none option.

a74nh replied:
Updated to the above. The CSR will need an update too.

theRealAph reviewed (approved):
Looks fine. I've done some basic testing on Graviton 3, which all seems to work.

a74nh commented:
Coming back to this work for 2021....

I've just done a complete run of all the jtreg tests on this patch with both PAC on and PAC off on a PAC enabled machine.
With PAC off I saw no regressions. With PAC on I saw ~250 regressions - mostly Shenendoah tests, but with a few ZGC and serviceability tests too. Is it worth holding this patch up for those fixes?

In addition the CSR still needs a review: https://bugs.openjdk.java.net/browse/JDK-8277543

adinn commented:
> Is it worth holding this patch up for those fixes?

Probably. It depends on exactly what is failing. Can you provide more info?

a74nh commented:
>Probably. It depends on exactly what is failing. Can you provide more info?

Looks like they are all segfaults, which is exactly what I'd expect if a frame was missing a sign or auth. I'm fairly confident it's the same handful of issues over and over. I'll get onto debugging them a little more.

adinn commented:
> Looks like they are all segfaults, which is exactly what I'd expect if a frame was missing a sign or auth. I'm fairly confident it's the same handful of issues over and over. I'll get onto debugging them a little more.

I was hoping you were going to say that ... :-) Let's see how it looks in the debugger.

a74nh commented:
The new commits fix all the GC issues and fixes up the jvmci test.

There are still a few nsk tests that need fixing. I'm looking at those now.

adinn commented on src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp:
```diff
@@ -832,6 +832,7 @@ void TemplateInterpreterGenerator::generate_fixed_frame(bool native_call) {
   __ ldr(rcpool, Address(rcpool, ConstantPool::cache_offset_in_bytes()));
   __ stp(rlocals, rcpool, Address(sp, 2 * wordSize));
 
+  __ protect_return_address();
```
Most of the changes to fix the test look fairly self-explanatory but I don't really understand why you relocated call to protect_return-_address from its previous location at line 801. Can you explain why it has been moved?

a74nh replied:
I originally moved it as part of debugging (a GC load_at occurs during the load_mirror).
Once all the GC changes went in (all the enter_subframe calls), this change was no longer required.
Then, when I came to change it back, I realised it made more sense in the new place. The protect is now directly before the storing of lr to the stack. That's logically a better place and should make the assembler easier to read.

adinn replied:
Ok, I thought the new place looked better but was not clear why it had not been there in the first place. Thanks for clarifying.

adinn commented:
These test fixes all look like they are doing the right thing and  are localized to systems that have PAC_RET support.

It was a bit of a surprise to realize that double-protecting was a faux pas and, in consequence, that there there were places where a protected address was being encountered where a raw one was needed. It makes me wonder if there are other places where that might happen lying dormant which just happen not to have been found by any existing test. Do we have any way of systematically excluding that possibility (I supsect the answer is no but I have to ask).

a74nh commented:
> These test fixes all look like they are doing the right thing and are localized to systems that have PAC_RET support.
> 
> It was a bit of a surprise to realize that double-protecting was a faux pas and, in consequence, that there there were places where a protected address was being encountered where a raw one was needed. It makes me wonder if there are other places where that might happen lying dormant which just happen not to have been found by any existing test. Do we have any way of systematically excluding that possibility (I supsect the answer is no but I have to ask).


check_return_address() is being a huge help here. Before signing, the check is used to confirm it's a raw value. Meaning we get a segfault at the point of the second sign. It's a fairly straightforward process to then figure out that issue.

(Without the check, the segfault would occur on return after the authentication. Which could be a long way away from the signing. From experience, that's not a fun debug journey.)

There might still be some places where that code path is currently not executed today by any of the tests. There's no obvious way to catch those issues.

Once this is in, getting some regular testing will be fairly crucial to make sure it doesn't rot.

adinn commented:
> check_return_address() is being a huge help here

Yes, I appreciate that rejecting double signing is a good model.The surprise was solely a reflection of my limited understanding.

> There's no obvious way to catch those issues.

Ok, well in that case I guess I'm all right for this to be pushed as is. However, the other reviewers really need to give a nod before you proceed.

> Once this is in, getting some regular testing will be fairly crucial to make sure it doesn't rot.

I am hoping we will get that *informally* as Java developers get their hands on shiny new M1 macs and start using them in anger. However, I agree that having regular test runs for mac-aarch64 is a necessity.

theRealAph commented:
> > > > Am I right is saying that for Macos, all generated code is remapped RO before execution?
> > > 
> > > 
> > > Ah, no, it seems the code cache is not RWX all the time as far as Java threads are concerned. The Macos/AArch64 code is strategically calling pthread_jit_write_protect_np at Java <-> JVM transition points.
> > 
> > 
> > And this requires magic kernel support. I did mention it to a kernel engineer who wasn't very impressed, but I think it's pretty cool.
> 
> It's possible to emulate this to some extent with memory protection keys on POWER and (recent) x86. See `pkey_alloc`.

I don't think this does exactly what we need, because (at least according to the docs) it does it for the whole process, not just the jit threads. Unless I've read the docs wrongly.

theRealAph reviewed (changes requested):
I've reverted my approval pending my question which wasn't resolved.

a74nh commented:
Can anyone give the CSR a review please? It's blocked on having a hotspot engineer review it.

https://bugs.openjdk.java.net/browse/JDK-8277543

dholmes-ora commented on src/hotspot/cpu/aarch64/globals_aarch64.hpp:
```diff
@@ -118,7 +118,9 @@ define_pd_global(intx, InlineSmallCode,          1000);
-          range(1, 99)
+          range(1, 99)                                                  \
+  product(ccstr, UseBranchProtection, "none",                           \
+          "Branch Protection to use: none,standard,pac-ret")            \
```
Nit: spaces after commas please.

dholmes-ora commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -5269,3 +5279,106 @@ void MacroAssembler::spin_wait() {
+
+// Stack frame creation/removal
+
+void MacroAssembler::enter()
+{
```
Style nit: opening braces go on the same line as the function declaration.

dholmes-ora commented on src/hotspot/cpu/aarch64/vm_version_aarch64.cpp:
```diff
@@ -407,6 +408,29 @@ void VM_Version::initialize() {
+    _rop_protection = false;
+    // Enable PAC if this code has been built with branch-protection and the CPU/OS supports it.
+#ifdef __ARM_FEATURE_PAC_DEFAULT
+    if (_features & CPU_PACA) {
```
Style nit: no implicit booleans - expand as "if ( A & B != 0)"

theRealAph replied:
Oh yuck, really? This is my punishment for not paying attention to the style guide dicussions.

dholmes-ora commented on src/hotspot/share/gc/shared/barrierSetNMethod.cpp:
```diff
@@ -54,6 +55,7 @@ int BarrierSetNMethod::nmethod_stub_entry_barrier(address* return_address_ptr) {
   MACOS_AARCH64_ONLY(ThreadWXEnable wx(WXWrite, Thread::current()));
 
   address return_address = *return_address_ptr;
+  AARCH64_ONLY(return_address=pauth_strip_pointer(return_address));
```
Style nit: spaces around binary operators please. There are a couple of occurrences.

dholmes-ora commented on src/hotspot/share/runtime/frame.cpp:
```diff
@@ -1110,6 +1110,9 @@ void frame::verify(const RegisterMap* map) const {
 
 #ifdef ASSERT
 bool frame::verify_return_pc(address x) {
+  AARCH64_ONLY(if (!pauth_ptr_is_raw(x)) {
+    return false;
+  })
```
Style nit: Use ifdef for multi-line code blocks.

dholmes-ora commented on src/hotspot/cpu/aarch64/vm_version_aarch64.cpp:
```diff
@@ -407,6 +408,29 @@ void VM_Version::initialize() {
+    }
+#endif
+  } else if (strcmp(UseBranchProtection, "pac-ret") == 0) {
+    _rop_protection = true;
+#ifdef __ARM_FEATURE_PAC_DEFAULT
+    if (!(_features & CPU_PACA)) {
+      warning("UseROPProtection specified, but not supported on this CPU.");
+    }
+#else
+    warning("UseROPProtection specified, but not supported in the VM.");
+#endif
```
If we issue these warnings should `_rop_protection` still be set true?

a74nh replied:
As per this conversation: https://github.com/openjdk/jdk/pull/6334#discussion_r791722292

The idea was, the user is explicitly asking for asking for pac-ret so we should honour that. Whereas standard would only enable what is supported for that system.

dholmes-ora replied:
But we can't honour that because it is not supported. Further, the suggestion in the referenced discussion seemed to be based on the assumption that doing so would be harmless because it is NOP based, but you have indicated that may not be the case and so it may actually lead to a crash!

a74nh replied:
Before I change anything - @theRealAph you had an opinion here too...

theRealAph replied:
Given that the implementation has now changed so much that it's no longer NOP based, I'll go with @dholmes-ora .
One other thing, though: it might be better to say here "but this VM was built without ROP-protection support." That's more informative, IMO.

a74nh replied:
Ok, I'll fix up as suggested.

The beginning part of that message needs fixing too - UseROPProtection is no longer the name of the flag. I'll switch to:
"ROP-protection specified, but this VM was built without ROP-protection support."

a74nh replied:
And this change will keep ROP protection enabled if we fall into the "this VM was built without ROP-protection support.". In that case we'll be protecting generated code, but the VM itself won't be protected. This will run without crashing.

theRealAph replied:
> And this change will keep ROP protection enabled if we fall into the "this VM was built without ROP-protection support.". In that case we'll be protecting generated code, but the VM itself won't be protected. This will run without crashing.

That's perfect.

dholmes-ora replied:
Okay, that makes sense. Thanks.

dholmes-ora reviewed (commented):
A few stylistic nits below. And one query.

Thanks,
David

a74nh commented:
As mentioned on the CSR, the JEP is being dropped - unless anyone has any objections.
JDK-8277204 will become a normal RFE.

theRealAph commented:
> As mentioned on the CSR, the JEP is being dropped - unless anyone has any objections. JDK-8277204 will become a normal RFE.

Good decision.

a74nh commented:
As requested in the RFE, added the new flag to the man page. Also updated the building.md instructions.

However, I'm not sure how to add to the release notes - I can't find any files or a process.

a74nh commented:
> However, I'm not sure how to add to the release notes - I can't find any files or a process.

Ok, This part I understand now :)

theRealAph commented on doc/building.md:
```diff
@@ -135,6 +135,13 @@ space is required.
+#### Branch Protection
+
+In order to use Branch Protection features in the VM, `--enable-branch-protection`
+must be provided. This requires compiler support (GCC 9.1.0+ or Clang 10+). The
```
```suggestion
must be used. This option requires C++ compiler support (GCC 9.1.0+ or Clang 10+). The
```

theRealAph commented on doc/building.md:
```diff
@@ -135,6 +135,13 @@ space is required.
+In order to use Branch Protection features in the VM, `--enable-branch-protection`
+must be provided. This requires compiler support (GCC 9.1.0+ or Clang 10+). The
+resulting build can be run on both machines with and without support for branch
+protection in hardware. This is only supported for Linux targets.
```
```suggestion
protection in hardware. Branch Protection is only supported for Linux targets.
```

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -5269,3 +5279,103 @@ void MacroAssembler::spin_wait() {
+
+// Create an additional frame for a function.
+void MacroAssembler::enter_subframe() {
+  // Addresses can only be signed once, so strip it first. PAC safe because the value is not
```
This needs a more descriptive name. `enter_and_sign()` ?

theRealAph replied:
Because enter always enters a subframe. That's what it's for.

a74nh replied:
enter_nested() ?
enter_inner() ?

theRealAph replied:
Tell you what, first put a comment here that says when it should (and therefore, should not) be used. Once it's clear exactly what this is for, thinking of a name maight be easier.

a74nh replied:
How about extending the existing enter() function: 

// Enter a new stack frame for the current method.
// nested:     Indicates a frame has already been entered (and not left) for the current method. 
void MacroAssembler::enter(bool nested=false) {
   if (nested) strip()
   protect()
   stp()
   mov()
}

This would add an additional bool check for every call of enter() - that's at code generation time, so probably not an issue.

theRealAph replied:
So, `nested` is true iff we are, say, pushing an extra frame for a runtime call in the middle of generated code, but for some mysterious reason the logic is inline instead of being implemented in the obvious way as a stub.

Please do this as:

` MacroAssembler::enter(bool strip_return_address=false)`

and I'll be happy. Please make sure that all calls are commented, as in

`__ enter(/*strip_return_address*/true);`

and I'll be happy.

a74nh replied:
Just about to resolve this ... then spotted the "make sure that all calls are commented". Will fix up.

theRealAph commented on src/hotspot/cpu/aarch64/frame_aarch64.cpp:
```diff
@@ -268,17 +268,18 @@ bool frame::safe_for_sender(JavaThread *thread) {
   if (TracePcPatching) {
     tty->print_cr("patch_pc at address " INTPTR_FORMAT " [" INTPTR_FORMAT " -> " INTPTR_FORMAT "]",
-                  p2i(pc_addr), p2i(*pc_addr), p2i(pc));
+                  p2i(pc_addr), p2i(*pc_addr), p2i(signed_pc));
```
Let's see both pc and signed pc here.

theRealAph replied:
> Let's see both pc and signed pc here, if they are different.

a74nh replied:
Are you sure? At the moment with PAC we get:

patch_pc at address 0x0000fffff58edf98 [0x0068ffffed17b5fc -> 0x00abffffed17b7f8]

With both signed and unsigned you'd have:

patch_pc at address 0x0000fffff58edf98 [0x0068ffffed17b5fc (0x0000ffffed17b5fc) -> 0x00abffffed17b7f8 (0x0000ffffed17b7f8)]

I prefer the first - it's shorter and you can infer the address from the signed version. Happy to go with the longer version if you think the shorter version is confusing.

theRealAph replied:
You've been looking at PAC-signed addresses for a long time. 
Let's see "at address [prev true dest -> new true dest] [signed prev signed dest -> new signed dest]", but only show the signed dests if they're different. So it appears as 
`patch_pc at address 0x0000fffff58edf98 [0x0068ffffed17b5fc -> 0x00abffffed17b7f8] [signed 0x0000ffffed17b5fc ->  0x0000ffffed17b7f8]` .

a74nh replied:
ok, that looks better than my longer version, I'll go with that

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -5269,3 +5279,103 @@ void MacroAssembler::spin_wait() {
+// frame for the current function.
+// Uses the FP from the start of the function as the modifier - which is stored at the address of
+// the current FP.
+//
```
Is it? C2 uses FP as a scratch register. I guess we know that this is never used in C2-generated code? I'm tempted to put an assertion here, just in case. Or does it not matter?

theRealAph replied:
If `UseROPProtection` is on, is there any reason not to set `PreserveFramePointer`, and assert here that it is set? It is a crucial assumption, so let's assert it.

a74nh replied:
PreserveFramePointer is doing some additional stuff. I'll give it a test to make sure everything still works with PreserveFramePointer fully set. It would make things easier just to force it set with rop protection on.

theRealAph replied:
> PreserveFramePointer is doing some additional stuff. I'll give it a test to make sure everything still works with PreserveFramePointer fully set. It would make things easier just to force it set with rop protection on.

Using PreserveFramePointer greatly simplifies the testing matrix, and has little adverse performance impact beyond disallowing C2 from allocating FP as a scratch register. It also simplifies this patch, which would be a very Good Thing. Let's do it.

a74nh replied:
Doing this caused 7 failures across a full jtreg run, namely:

serviceability/sa/ClhsdbFindPC.java#xcomp-core
vmTestbase/jit/misctests/fpustack/GraphApplet.java
vmTestbase/nsk/jdi/MonitorWaitRequest/MonitorWaitRequest001/TestDescription.java
vmTestbase/nsk/jdi/MonitorWaitedRequest/MonitorWaitedRequest001/TestDescription.java
vmTestbase/nsk/jdwp/ThreadReference/ForceEarlyReturn/forceEarlyReturn002/forceEarlyReturn002.java
vmTestbase/nsk/jdwp/ThreadReference/OwnedMonitorsStackDepthInfo/ownedMonitorsStackDepthInfo002/ownedMonitorsStackDepthInfo002.java
vmTestbase/nsk/jvmti/RedefineClasses/StressRedefine/TestDescription.java

....I'll investigate.

theRealAph replied:
> Doing this caused 7 failures across a full jtreg run, namely:

I'm glad we caught that.

theRealAph replied:
Status? Is branch protection really incompatible with PreserveFramePointer?

theRealAph commented on src/hotspot/cpu/aarch64/vm_version_aarch64.cpp:
```diff
@@ -433,6 +433,11 @@ void VM_Version::initialize() {
+  // The frame pointer must be preserved for ROP protection.
+  if (_rop_protection == true) {
+    PreserveFramePointer = true;
+  }
```
You need an error message for -PreserveFramePointer +UseROPProtection.

theRealAph reviewed (commented):
This is looking pretty nice now. With the check for -XX:-UseFramePointer argument consistency we're done.

a74nh commented:
> This is looking pretty nice now. With the check for -XX:-UseFramePointer argument consistency we're done.

Excellent! I'm away all next week, so will add the check when I get back.

a74nh commented:
I did another full jteg run, and everything looks fine.
Think that's all the review comments resolved now too.

a74nh commented:
Any more comments? Otherwise I'll integrate later

a74nh commented:
/integrate

adinn commented:
/sponsor

a74nh commented:
Yes, we spotted this today too. https://bugs.openjdk.java.net/browse/JDK-8282392

My initial thought was that I needed to add a pauth header file with stub functions to linux_zero/. Which does feel a little awkward.

AARCH64_PORT_ONLY does sound like a better option.

Thankfully there is no need for full pac support in zero too.... :)

dougxc commented on src/hotspot/cpu/aarch64/vm_version_aarch64.cpp:
```diff
@@ -409,6 +410,39 @@ void VM_Version::initialize() {
     UsePopCountInstruction = true;
   }
 
+  if (UseBranchProtection == nullptr || strcmp(UseBranchProtection, "none") == 0) {
```
My understanding is that `UseBranchProtection` requires JIT compiler integration. This is currently missing in Graal. The logic that decides if it should be enabled by default needs to take JVMCI into consideration, much like the call to `JVMCIGlobals::check_jvmci_supported_gc` determines if JVMCI supports the configured GC.

Thanks to @gilles-duboscq for alerting me to this change.

shqking replied:
Thanks for reporting this. @dougxc 

Yes. I encountered several jtreg failures with Graal, which is built with PAC-RET enabled JDK.

As I see it, the straightforward fix is to disable JVMCI if `_rop_protection` is parsed as "true" finally, since Graal doesn't supports PAC-RET currently. 
That is similar to the way `JVMCIGlobals::check_jvmci_supported_gc` does, i.e. if one GC except `G1/Serial/Parallel` is configured, we disable JVMCI.
WDYT

dougxc replied:
The problem with `JVMCIGlobals::check_jvmci_supported_gc` is that it does not give the compiler a chance to specify whether it supports a given GC.
Instead, we want a JVMCI upcall like `jdk.vm.ci.hotspot.HotSpotJVMCIRuntime.isGCSupported(int)`.

shqking replied:
Thanks for your reply. But I'm afraid I'm still a bit confused.

1) flag `_rop_protection` is set inside function `VM_Version::initialize()`. The execution flow is `create_vm -> init_globals -> VM_Version_init -> VM_Version::initialize()`.
2) I suppose compilers (including JVMCI compiler) are initialized after `init_global()` function. See https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/threads.cpp#L685~L707. Hence I suppose we cannot call JVMCI methods before this site, right?

If so, I'm afraid it's a bit late for VM to ask the compiler whether it supports PAC-RET after the compiler initialization, and to disable PAC-RET if the JVMCI compiler doesn't support it, mainly because some stubs are generated already between `init_global()` and compiler initialization, and in these stubs, flag `_rop_protection` is used to generate PAC-related code, e.g., `SharedRuntime::generate_stubs() -> generate_deopt_blob() -> protect_return_address()`.

Please correct me if there is something I misunderstood. Thanks.

dougxc replied:
Ah ok, I did not pay attention to the VM lifecycle constraints in setting this flag. We could do the JVMCI check in `CompileBroker::compilation_init_phase2` but that would mean forcing eager initialization of JVMCI which, in turn, would impact VM startup.
For now, we will just add a check into Graal that `UseBranchProtection` is not set, exiting the VM if it is. At the same time, we will work on adding support for this.

a74nh replied:
>For now, we will just add a check into Graal that UseBranchProtection is not set, exiting the VM if it is

that sounds a better solution to me instead of:

> disable JVMCI if _rop_protection is parsed as "true"
